tests/components/zha/test_discover.py

```diff
@@ -30,45 +66,298 @@
+        disc.PROBE.discover_by_device_type(ep_channels)
+    if hit:
+        assert get_entity_mock.call_count == 1
+        assert ep_channels.claim_channels.call_count == 1
+        assert ep_channels.claim_channels.call_args[0][0] is mock.sentinel.claimed
+        assert ep_channels.async_new_entity.call_count == 1
+        assert ep_channels.async_new_entity.call_args[0][0] == component
+        assert ep_channels.async_new_entity.call_args[0][1] == mock.sentinel.entity_cls
 
-        entity_ids = hass.states.async_entity_ids()
-        await hass.async_block_till_done()
-        zha_entities = {
-            ent for ent in entity_ids if ent.split(".")[0] in zha_const.COMPONENTS
-        }
-
-        zha_gateway = get_zha_gateway(hass)
-        zha_dev = zha_gateway.get_device(zigpy_device.ieee)
-        event_channels = {  # pylint: disable=protected-access
-            ch.id for ch in zha_dev._relay_channels.values()
-        }
-
-        assert zha_entities == set(device["entities"])
-        assert event_channels == set(device["event_channels"])
-
-        entity_map = device["entity_map"]
-        for calls in _dispatch.call_args_list:
-            discovery_info = calls[0][2]
-            unique_id = discovery_info["unique_id"]
-            channels = discovery_info["channels"]
-            component = discovery_info["component"]
-            key = (component, unique_id)
-            entity_id = entity_registry.async_get_entity_id(component, "zha", unique_id)
-
-            assert key in entity_map
-            assert entity_id is not None
-            no_tail_id = NO_TAIL_ID.sub("", entity_map[key]["entity_id"])
-            assert entity_id.startswith(no_tail_id)
-            assert set([ch.name for ch in channels]) == set(entity_map[key]["channels"])
+
+def test_discover_by_device_type_override():
+    """Test entity discovery by device type overriding."""
+
+    ep_channels = mock.MagicMock(spec_set=zha_channels.ChannelPool)
+    ep_mock = mock.PropertyMock()
+    ep_mock.return_value.profile_id = 0x0104
+    ep_mock.return_value.device_type = 0x0100
+    type(ep_channels).endpoint = ep_mock
+
+    overrides = {ep_channels.unique_id: {"type": zha_const.SWITCH}}
+    get_entity_mock = mock.MagicMock(
+        return_value=(mock.sentinel.entity_cls, mock.sentinel.claimed)
+    )
+    with mock.patch(
+        "homeassistant.components.zha.core.registries.ZHA_ENTITIES.get_entity",
+        get_entity_mock,
+    ):
+        with mock.patch.dict(disc.PROBE._device_configs, overrides, clear=True):
+            disc.PROBE.discover_by_device_type(ep_channels)
+            assert get_entity_mock.call_count == 1
+            assert ep_channels.claim_channels.call_count == 1
+            assert ep_channels.claim_channels.call_args[0][0] is mock.sentinel.claimed
+            assert ep_channels.async_new_entity.call_count == 1
+            assert ep_channels.async_new_entity.call_args[0][0] == zha_const.SWITCH
+            assert (
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5051600)